File: Treachery-iOS/Treachery-iOS/Models/Game.swift
//
//  Game.swift
//  Treachery-iOS
//
//  Created by Luke Solomon on 3/12/26.
//

import Foundation

enum GameState: String, Codable {
    case waiting
    case inProgress = "in_progress"
    case finished
}

struct Game: Codable, Identifiable, Hashable {
    let id: String
    let code: String
    let hostId: String
    var state: GameState
    let maxPlayers: Int
    let startingLife: Int
    var winningTeam: String?
    var gameMode: GameMode
    var playerIds: [String]
    let createdAt: Date
    var lastActivityAt: Date?
    var planechase: PlanechaseState?

    enum CodingKeys: String, CodingKey {
        case id
        case code
        case hostId = "host_id"
        case state
        case gameMode = "game_mode"
        case maxPlayers = "max_players"
        case startingLife = "starting_life"
        case winningTeam = "winning_team"
        case playerIds = "player_ids"
        case createdAt = "created_at"
        case lastActivityAt = "last_activity_at"
        case planechase
    }

    // Custom decoder: defaults playerIds to [] for documents created before
    // this field existed, preventing decoding crashes on legacy data.
    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        id = try container.decode(String.self, forKey: .id)
        code = try container.decode(String.self, forKey: .code)
        hostId = try container.decode(String.self, forKey: .hostId)
        state = try container.decode(GameState.self, forKey: .state)
        maxPlayers = try container.decode(Int.self, forKey: .maxPlayers)
        startingLife = try container.decode(Int.self, forKey: .startingLife)
        gameMode = try container.decodeIfPresent(GameMode.self, forKey: .gameMode) ?? .treachery
        winningTeam = try container.decodeIfPresent(String.self, forKey: .winningTeam)
        playerIds = try container.decodeIfPresent([String].self, forKey: .playerIds) ?? []
        createdAt = try container.decode(Date.self, forKey: .createdAt)
        lastActivityAt = try container.decodeIfPresent(Date.self, forKey: .lastActivityAt)
        planechase = try container.decodeIfPresent(PlanechaseState.self, forKey: .planechase)
    }

    init(
        id: String,
        code: String,
        hostId: String,
        state: GameState,
        gameMode: GameMode = .treachery,
        maxPlayers: Int,
        startingLife: Int,
        winningTeam: String?,
        playerIds: [String],
        createdAt: Date,
        lastActivityAt: Date? = nil,
        planechase: PlanechaseState? = nil
    ) {
        self.id = id
        self.code = code
        self.hostId = hostId
        self.state = state
        self.gameMode = gameMode
        self.maxPlayers = maxPlayers
        self.startingLife = startingLife
        self.winningTeam = winningTeam
        self.playerIds = playerIds
        self.createdAt = createdAt
        self.lastActivityAt = lastActivityAt
        self.planechase = planechase
    }
}

// MARK: - Game Errors

enum GameError: LocalizedError {
    case codeGenerationFailed
    case gameFull
    case gameNotFound
    case gameAlreadyStarted
    case cardAssignmentFailed

    var errorDescription: String? {
        switch self {
        case .codeGenerationFailed: return "Could not generate a unique game code. Please try again."
        case .gameFull: return "This game is full."
        case .gameNotFound: return "No game found with that code."
        case .gameAlreadyStarted: return "This game has already started."
        case .cardAssignmentFailed: return "Could not assign identity cards. Please try again."
        }
    }
}

File: TreacheryAndroid/app/src/main/java/com/solomon/treachery/model/Game.kt
package com.solomon.treachery.model

import com.google.firebase.Timestamp

data class Game(
    val id: String = "",
    val code: String = "",
    val hostId: String = "",
    val state: GameState = GameState.WAITING,
    val maxPlayers: Int = 8,
    val startingLife: Int = 40,
    val winningTeam: String? = null,
    val gameMode: GameMode = GameMode.TREACHERY,
    val playerIds: List<String> = emptyList(),
    val createdAt: Timestamp = Timestamp.now(),
    val lastActivityAt: Timestamp? = null,
    val planechase: PlanechaseState? = null,
    val winnerUserIds: List<String> = emptyList()
) {
    fun toMap(): Map<String, Any?> = mapOf(
        "code" to code,
        "host_id" to hostId,
        "state" to state.value,
        "max_players" to maxPlayers,
        "starting_life" to startingLife,
        "winning_team" to winningTeam,
        "game_mode" to gameMode.value,
        "player_ids" to playerIds,
        "created_at" to createdAt,
        "last_activity_at" to lastActivityAt,
        "planechase" to planechase?.toMap(),
        "winner_user_ids" to winnerUserIds
    )

    companion object {
        fun fromMap(id: String, data: Map<String, Any?>): Game = Game(
            id = id,
            code = data["code"] as? String ?: "",
            hostId = data["host_id"] as? String ?: "",
            state = GameState.fromValue(data["state"] as? String ?: "") ?: GameState.WAITING,
            maxPlayers = (data["max_players"] as? Number)?.toInt() ?: 8,
            startingLife = (data["starting_life"] as? Number)?.toInt() ?: 40,
            winningTeam = data["winning_team"] as? String,
            gameMode = GameMode.fromValue(data["game_mode"] as? String ?: "") ?: GameMode.TREACHERY,
            playerIds = (data["player_ids"] as? List<*>)?.filterIsInstance<String>() ?: emptyList(),
            createdAt = data["created_at"] as? Timestamp ?: Timestamp.now(),
            lastActivityAt = data["last_activity_at"] as? Timestamp,
            planechase = (data["planechase"] as? Map<*, *>)?.let {
                @Suppress("UNCHECKED_CAST")
                PlanechaseState.fromMap(it as Map<String, Any?>)
            },
            winnerUserIds = (data["winner_user_ids"] as? List<*>)?.filterIsInstance<String>() ?: emptyList()
        )
    }
}
